File: Python/2-Medium/00019._删除链表的倒数第_N_个结点/00019._删除链表的倒数第_N_个结点.py
class ListNode:
    def __init__(self, val=0, next=None):
        """链表节点的初始化函数
        
        Args:
            val (int): 节点值，默认为0
            next (ListNode): 指向下一个节点的指针，默认为None
        """
        self.val = val
        self.next = next
def removeNthFromEnd(head: ListNode, n: int) -> ListNode:
    """删除链表的倒数第N个节点并返回链表头节点
    
    Args:
        head (ListNode): 链表的头节点
        n (int): 需要删除的倒数第N个节点
    
    Returns:
        ListNode: 删除节点后的链表头节点
    """
    dummy = ListNode(0, head)  # 创建一个哑节点，指向头节点，简化边界情况处理
    fast = slow = dummy  # 初始化快慢指针指向哑节点
    
    # 快指针先前移n+1步
    for _ in range(n + 1):
        fast = fast.next
    
    # 快慢指针同时移动，直到快指针到达链表末尾
    while fast:
        fast = fast.next
        slow = slow.next
    
    # 删除倒数第N个节点
    slow.next = slow.next.next
    
    return dummy.next  # 返回哑节点的下一个节点，即新的头节点


# 辅助函数：构建链表
def buildLinkedList(elements: list) -> ListNode:
    dummy = ListNode(0)
    current = dummy
    for element in elements:
        current.next = ListNode(element)
        current = current.next
    return dummy.next

# 辅助函数：打印链表
def printLinkedList(head: ListNode) -> None:
    elements = []
    while head:
        elements.append(str(head.val))
        head = head.next
    print("->".join(elements))

# 构建链表并测试
head = buildLinkedList([1, 2, 3, 4, 5])
n = 2
head = removeNthFromEnd(head, n)
printLinkedList(head)  # 预期输出：1->2->3->5

File: C++/2-Medium/00019._删除链表的倒数第_N_个结点/00019._删除链表的倒数第_N_个结点.cpp
#include <iostream>
using namespace std;

// 定义链表节点结构体
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// 删除链表的倒数第N个节点并返回头节点
ListNode* removeNthFromEnd(ListNode* head, int n) {
    // 创建一个虚拟头节点，方便处理边界情况
    ListNode* dummy = new ListNode(0);
    dummy->next = head;
    ListNode *fast = dummy, *slow = dummy;
    
    // fast指针先前移动n+1步
    for(int i = 0; i < n + 1; i++) {
        fast = fast->next;
    }
    
    // 同时移动fast和slow指针
    while(fast != NULL) {
        fast = fast->next;
        slow = slow->next;
    }
    
    // 删除slow指针的下一个节点
    ListNode* toDelete = slow->next;
    slow->next = slow->next->next;
    delete toDelete; // 释放内存
    
    // 返回更新后的链表头节点
    ListNode* newHead = dummy->next;
    delete dummy; // 释放虚拟头节点的内存
    return newHead;
}

// 主函数，用于演示删除操作
int main() {
    // 创建示例链表 1->2->3->4->5
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);
    
    int n = 2; // 设置需要删除的倒数第N个节点
    ListNode* result = removeNthFromEnd(head, n);
    
    // 打印结果链表
    while(result != NULL) {
        cout << result->val << "->";
        result = result->next;
    }
    cout << "NULL" << endl;
    
    return 0;
}